src/models/poll.ts

```diff
@@ -0,0 +1,178 @@
+    [PollEvent.Responses]: (responses: Relations) => void;
+};
+
+const filterResponseRelations = (
+    relationEvents: MatrixEvent[],
+    pollEndTimestamp: number,
+): {
+    responseEvents: MatrixEvent[];
+} => {
+    const responseEvents = relationEvents.filter((event) => {
+        if (event.isDecryptionFailure()) {
+            // @TODO(kerrya) PSG-1023 track and return these
+            return;
+        }
+        return (
+            M_POLL_RESPONSE.matches(event.getType()) &&
+            // From MSC3381:
+            // "Votes sent on or before the end event's timestamp are valid votes"
+            event.getTs() <= pollEndTimestamp
+        );
+    });
+
+    return { responseEvents };
+};
+
+// https://github.com/matrix-org/matrix-spec-proposals/pull/3672
+export class Poll extends TypedEventEmitter<Exclude<PollEvent, PollEvent.New>, PollEventHandlerMap> {
+    public readonly roomId: string;
+    private pollEvent: PollStartEvent | undefined;
+    private fetchingResponsesPromise: null | Promise<void> = null;
+    private responses: null | Relations = null;
+    private endEvent: MatrixEvent | undefined;
+
+    public constructor(private rootEvent: MatrixEvent, private matrixClient: MatrixClient) {
+        super();
+        this.roomId = this.rootEvent.getRoomId()!;
+        this.setPollStartEvent(this.rootEvent);
+    }
+
+    public get pollId(): string {
+        return this.rootEvent.getId()!;
+    }
+
+    public get isEnded(): boolean {
+        return !!this.endEvent;
+    }
+
+    public setPollStartEvent(event: MatrixEvent): void {
+        this.pollEvent = event.unstableExtensibleEvent as PollStartEvent;
+    }
+
+    public getPollStartEvent(): PollStartEvent {
+        return this.pollEvent!;
+    }
+
+    public async getResponses(): Promise<Relations> {
+        // if we have already fetched the responses
+        // just return them
+        if (this.responses) {
+            return this.responses;
+        }
+        if (!this.fetchingResponsesPromise) {
+            this.fetchingResponsesPromise = this.fetchResponses();
+        }
+        await this.fetchingResponsesPromise;
+        return this.responses!;
+    }
+
+    /**
+     *
+     * @param event - event with a relation to the rootEvent
+     * @returns void
+     */
+    public onNewRelation(event: MatrixEvent): void {
+        if (M_POLL_END.matches(event.getType())) {
+            this.endEvent = event;
+            this.refilterResponsesOnEnd();
+            this.emit(PollEvent.End);
+        }
+
+        // wait for poll responses to be initialised
+        if (!this.responses) {
+            return;
+        }
+
+        const pollEndTimestamp = this.endEvent?.getTs() || Number.MAX_SAFE_INTEGER;
+        const { responseEvents } = filterResponseRelations([event], pollEndTimestamp);
+
+        if (responseEvents.length) {
+            responseEvents.forEach((event) => {
+                this.responses!.addEvent(event);
```

Comment:
Is the `!` necessary here because of this check https://github.com/matrix-org/matrix-js-sdk/pull/3036/files#diff-6d5fabe09478e35a518f33e36f448a7e7c3e224ffa9a4a726fe8ccfeefd7b4f0R117 ?

Reply:
Yeah, I also thought ts would be able to infer that. Even wrapping this block in `if (this.responses) {...}` doesn't help

Reply:
:man_shrugging: